server/aesdsocket.c

```diff
@@ -11,144 +21,245 @@
 #include <sys/socket.h>
 #include <signal.h>
 #include <syslog.h>
+#include <pthread.h>
+#include <time.h>
 
 #define SERVER_PORT 9000
 #define BUFFER_SIZE 1024
 
-int sockfd, connfd;
+int sockfd;
+pthread_mutex_t mutex;
+typedef struct thread_node {
+    pthread_t tid;
+    struct thread_node *next;
+} node_t;
 
+node_t *head = NULL;
+pthread_mutex_t file_write_mutex = PTHREAD_MUTEX_INITIALIZER;
+pthread_mutex_t thread_list_mutex = PTHREAD_MUTEX_INITIALIZER;
+
+// Signal handler to catch SIGINT and SIGTERM for graceful shutdown.
 void signal_handler(int signal) {
     if (signal == SIGINT || signal == SIGTERM) {
-        // Close the socket
         close(sockfd);
-        // Delete the file
         remove("/var/tmp/aesdsocketdata");
-        // Log the message
         syslog(LOG_INFO, "Caught signal, exiting");
-        // Close syslog
         closelog();
         exit(0);
+    } else if (signal == SIGALRM) {
+        char timestamp_str[100];
+        time_t current_time = time(NULL);
+        struct tm *local_time = localtime(&current_time);
+        strftime(timestamp_str, sizeof(timestamp_str), "timestamp:%a, %d %b %Y %H:%M:%S %z\n", local_time);
+        
+        pthread_mutex_lock(&mutex);
+        FILE *fp = fopen("/var/tmp/aesdsocketdata", "a");
+        if (fp == NULL) {
+            perror("fopen: Failed opening file.");
+        } else {
+            fprintf(fp, "%s", timestamp_str);
+            fclose(fp);
+        }
+        pthread_mutex_unlock(&mutex);
     }
 }
 
+void *connection_handler(void *socket_desc) {
+    int connfd = *(int *)socket_desc;
+    char *buffer = calloc(BUFFER_SIZE, sizeof(char));
+    
+    while (recv(connfd, buffer, BUFFER_SIZE, 0) > 0) {
+        pthread_mutex_lock(&mutex);
+        FILE *fp = fopen("/var/tmp/aesdsocketdata", "a");
+        if (fp == NULL) {
+            perror("fopen: Failed opening file.");
+            pthread_mutex_unlock(&mutex);
+            free(buffer);
+            close(connfd);
+            return NULL;
+        } else {
+            fprintf(fp, "%s", buffer);
```

Comment:
@tiba6275 what would happen if a packet was longer than `BUFFER_SIZE` with this design?  Would the packet potentially be split up by a packet received from another thread or a timestamp print?